src/h3lib/lib/iterators.c

```diff
@@ -0,0 +1,271 @@
+| res | base cell # | 1 | 2 | 3 | 4 | 5 | 6 | ... |
+|-----|-------------|---|---|---|---|---|---|-----|
+|   6 |          97 | 0 | 0 | 0 | 0 | 2 | 0 | ... |
+                                   \
+                                    first nonzero digit
+
+iterStepChild ->
+
+We won't skip the `1` in a previous digit when it
+is no longer the first nonzero digit.
+
+                            parent res      child res
+                           /               /
+| res | base cell # | 1 | 2 | 3 | 4 | 5 | 6 | ... |
+|-----|-------------|---|---|---|---|---|---|-----|
+|   6 |          97 | 0 | 0 | 0 | 0 | 2 | 1 | ... |
+                                   \
+                                    first nonzero digit
+ */
+
+/*
+Initialize a Iter_Child struct representing the sequence giving
+the children of cell `h` at resolution `childRes`.
+
+At any point in the iteration, starting once
+the struct is initialized, Iter_Child.h gives the current child.
+
+Also, Iter_Child.h == H3_NULL when all the children have been iterated
+through, or if the input to `iterInitParent` was invalid.
+ */
+Iter_Child iterInitParent(H3Index h, int childRes) {
+    Iter_Child it;
+
+    it.parentRes = H3_GET_RESOLUTION(h);
+
+    if (childRes < it.parentRes || childRes > MAX_H3_RES || h == H3_NULL) {
+        return _null_iter();
+    }
+
+    it.h = _zero_index_digits(h, it.parentRes + 1, childRes);
+    H3_SET_RESOLUTION(it.h, childRes);
+
+    if (H3_EXPORT(h3IsPentagon)(it.h))
+        // The first nonzero digit skips `1` for pentagons.
+        // The "fnz" moves to the left as we count up from the child resolution
+        // to the parent resolution.
+        it.fnz = childRes;
+    else
+        // if not a pentagon, we can ignore "first nonzero digit" logic
+        it.fnz = -1;
+
+    return it;
+}
+
+/*
+Step a Iter_Child to the next child cell.
+When the iteration is over, Iter_Child.h will be H3_NULL.
+Handles iterating through hexagon and pentagon cells.
+ */
+void iterStepChild(Iter_Child* it) {
+    // once h == H3_NULL, the iterator returns an infinite sequence of H3_NULL
+    if (it->h == H3_NULL) return;
+
+    int childRes = H3_GET_RESOLUTION(it->h);
+
+    _inc(it, childRes);
+
+    for (int i = childRes; i >= it->parentRes; i--) {
+        if (i == it->parentRes) {
+            // if we're modifying the parent resolution digit, then we're done
+            *it = _null_iter();
+            return;
+        }
+
+        if (i == it->fnz && _get(it, i) == 1) {
+            // Then we are iterating through the children of a pentagon cell.
+            // All children of a pentagon have the property that the first
+            // nonzero digit between the parent and child resolutions is
+            // not 1.
+            // I.e., we never see a sequence like 00001.
+            // Thus, we skip the `1` in this digit.
+            _inc(it, i);
+            it->fnz -= 1;
+            return;
+        }
+
+        if (_get(it, i) == 7) {
```

Comment:
Please use named Direction constants for INVALID_DIGIT, K_AXIS_DIGIT, etc.

Reply:
updated here!